eclair-core/src/main/scala/fr/acinq/eclair/payment/MultiPartPaymentHandler.scala

```diff
@@ -0,0 +1,143 @@
+
+  startWith(WAITING_FOR_HTLC, WaitingForHtlc(0 msat, Nil))
+
+  when(WAITING_FOR_HTLC) {
+    case Event(PaymentTimeout, d: WaitingForHtlc) =>
+      goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, wire.PaymentTimeout, d.parts)
+
+    case Event(MultiPartHtlc(totalAmount2, htlc), d: WaitingForHtlc) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      val pp = PartialPayment(htlc.amountMsat, htlc.channelId)
+      if (totalAmount != totalAmount2) {
+        log.warning(s"multi-part payment total amount mismatch: previously $totalAmount, now $totalAmount2")
+        goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, IncorrectOrUnknownPaymentDetails(totalAmount2, nodeParams.currentBlockHeight), PendingPayment(htlc.id, pp, sender) :: d.parts)
+      } else if (htlc.amountMsat + d.paidAmount >= totalAmount) {
+        goto(PAYMENT_SUCCEEDED) using PaymentSucceeded(htlc.amountMsat + d.paidAmount, PendingPayment(htlc.id, pp, sender) :: d.parts)
+      } else {
+        stay using d.copy(paidAmount = d.paidAmount + htlc.amountMsat, parts = PendingPayment(htlc.id, pp, sender) :: d.parts)
+      }
+  }
+
+  when(PAYMENT_SUCCEEDED) {
+    // A sender should not send us additional htlcs for that payment once we've already reached the total amount.
+    // However if that happens the only rational choice is to fulfill it, because the pre-image has been released so
+    // intermediate nodes will be able to fulfill that htlc anyway. This is a harmless spec violation.
+    case Event(MultiPartHtlc(_, htlc), _) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      log.info(s"received extraneous htlc for payment hash $paymentHash")
+      parent ! ExtraHtlcReceived(paymentHash, PendingPayment(htlc.id, PartialPayment(htlc.amountMsat, htlc.channelId), sender), None)
+      stay
+
+    case Event("ok", _) => stay
+  }
+
+  when(PAYMENT_FAILED) {
+    // If we receive htlcs after the multi-part payment has expired, we must fail them.
+    // The LocalPaymentHandler will create a new instance of MultiPartPaymentHandler to handle a new attempt.
+    case Event(MultiPartHtlc(_, htlc), PaymentFailed(_, failure, _)) =>
+      require(htlc.paymentHash == paymentHash, s"invalid payment hash (expected $paymentHash, received ${htlc.paymentHash}")
+      parent ! ExtraHtlcReceived(paymentHash, PendingPayment(htlc.id, PartialPayment(htlc.amountMsat, htlc.channelId), sender), Some(failure))
+      stay
+
+    case Event("ok", _) => stay
+  }
+
+  onTransition {
+    case _ -> PAYMENT_SUCCEEDED =>
+      cancelTimer(PaymentTimeout.toString)
+      nextStateData match {
+        case PaymentSucceeded(_, parts) =>
+          // We expect the parent actor to send us a PoisonPill after receiving this message.
+          parent ! MultiPartHtlcSucceeded(paymentHash, parts.reverse)
```

Comment:
I don't quite understand why you didn't send `MultiPartHtlcSucceeded`/`MultiPartHtlcFailed` right before leaving the `WAITING_FOR_HTLC` state:

```scala
if (totalAmount != totalAmount2) {
        log.warning(s"multi-part payment total amount mismatch: previously $totalAmount, now $totalAmount2")
        parent ! MultiPartHtlcFailed(paymentHash, failure, parts.reverse)
        goto(PAYMENT_FAILED) using PaymentFailed(d.paidAmount, IncorrectOrUnknownPaymentDetails(totalAmount2, nodeParams.currentBlockHeight), PendingPayment(htlc.id, pp, sender) :: d.parts)
      } else if (htlc.amountMsat + d.paidAmount >= totalAmount) {
        parent ! MultiPartHtlcSucceeded(paymentHash, parts.reverse)
        goto(PAYMENT_SUCCEEDED) using PaymentSucceeded(htlc.amountMsat + d.paidAmount, PendingPayment(htlc.id, pp, sender) :: d.parts)
      } else {
        stay using d.copy(paidAmount = d.paidAmount + htlc.amountMsat, parts = PendingPayment(htlc.id, pp, sender) :: d.parts)
      }
```

Reply:
Not sure if states `PAYMENT_SUCCEEDED`/`PAYMENT_FAILED` are really needed. I would just `stop()` the FSM with a `Normal`/`Failure` reason. Extra htlcs would be just failed, which isn't a bad thing I think?

Reply:
I think the state machine is easier to reason about (and avoid hidden race conditions) with the `PAYMENT_SUCCEEDED / PAYMENT_FAILED` states. It allows us to easily log these small spec violations to detect bugs in sender implementations.

That's also why I'm sending the response to the parent in the `onTransition`: it factors that into a single place which is simpler than having it in potentially multiple `if` branches if later commits update this FSM.

In the success case I think we really want to fulfill additional HTLCs, not fail them. They indicate a sender bug, but the only way we can refund the user is if we fulfilled the extra payments (otherwise an intermediate node will claim them but we can't refund because we didn't accept them).